stats.c

```diff
@@ -2020,34 +2023,31 @@ int replicate_regions(stats_t *stats, hts_itr_multi_t *iter, stats_info_t *info)
     return 0;
 }
 
-void init_group_id(stats_t *stats, const char *id)
+static void init_group_id(stats_t *stats, stats_info_t *info, const char *id)
 {
-#if 0
-    if ( !stats->sam_header->dict )
-        stats->sam_header->dict = sam_header_parse2(stats->sam_header->text);
-    void *iter = stats->sam_header->dict;
-    const char *key, *val;
-    int n = 0;
-    stats->rg_hash = khash_str2int_init();
-    while ( (iter = sam_header2key_val(iter, "RG","ID","SM", &key, &val)) )
-    {
-        if ( !strcmp(id,key) || (val && !strcmp(id,val)) )
-        {
-            khiter_t k = kh_get(kh_rg, stats->rg_hash, key);
-            if ( k != kh_end(stats->rg_hash) )
-                fprintf(stderr, "[init_group_id] The group ID not unique: \"%s\"\n", key);
-            int ret;
-            k = kh_put(kh_rg, stats->rg_hash, key, &ret);
-            kh_value(stats->rg_hash, k) = val;
-            n++;
+    stats->rg_hash = kh_init(rg);
+    if (!stats->rg_hash) return;
+    sam_hdr_t *hdr = info->sam_header;
+    const char *key;
+    kstring_t sm = KS_INITIALIZE;
+    int i, ret, nrg = sam_hdr_count_lines(hdr, "RG");
```

Comment:
You need to check for `nrg < 0` here, so otherwise malformed BAM or CRAM headers go mostly undetected, and you get a lot of zeros in the output as everything is filtered out.  Ideally this function should also return a value (or maybe call `error()`) so bad headers and memory problems don't get glossed over.

Reply:
Thanks. Done.

Reply:
Er, possibly not completely.  There's also the `if (ret == -1)` cases below...